File: backend/opencheck/batch.py
"""Batch / portfolio screening — a thin loop over the single-LEI pipeline.

Phase 164. Everything in OpenCheck has been one LEI at a time; the people
who most need it — a compliance officer with a counterparty list, a
journalist with a leak, an EITI secretariat with ninety licence-holders —
arrive with a list. This module is deliberately *not* a second pipeline:
each row is exactly ``routers.lookup._lookup_impl`` for that LEI (so it
shares the replay cache, the degradation channel, the GLEIF throttle and
every source adapter unchanged), and the batch is the loop, the cap and
the concurrency limit around it.

Two numbers govern the shape, and both are stated on the page rather than
hidden:

* **Cap of 20 rows** (``MAX_ROWS``). A cold anchor costs 4–6 GLEIF calls,
  so twenty rows ≈ 80–120 calls ≈ two minutes at the Phase 143 throttle
  (50/min). Fifty rows, the first draft's cap, would have been five.
* **Two pipelines in flight** (``OPENCHECK_BATCH_CONCURRENCY``). More than
  that and one batch monopolises the shared GLEIF budget that every human
  lookup queues behind.

Rows come back **as they finish, not in input order**, and a row that could
not be checked — GLEIF rate-limited past the Phase 162 max-wait, an unknown
LEI, an adapter fan-out that died — is reported as ``row_failed`` with
``degraded: true``, never silently dropped and never rendered as a clean
row. The verdict-sentence rule from Phase 132 holds for the table too:
OpenCheck does not grade companies, so rows are never ranked by severity.
"""

from __future__ import annotations

import asyncio
import csv
import io
import json
import re
from dataclasses import dataclass, field
from typing import Any, AsyncIterator

from fastapi import HTTPException

from . import identifiers
from .config import get_settings
from .gleif_throttle import GleifRateLimitedError

#: Hard cap on rows per batch. Stated on the page and in the 422 detail.
MAX_ROWS = 20

#: Tolerant paste: LEIs separated by newlines, commas, semicolons, spaces
#: or tabs, so a column copied out of a spreadsheet just works.
_SPLIT = re.compile(r"[\s,;]+")


@dataclass(frozen=True)
class RejectedToken:
    """A pasted token that will not be screened, and why — shown in place."""

    token: str
    reason: str

    def to_dict(self) -> dict[str, str]:
        return {"token": self.token, "reason": self.reason}


@dataclass
class ParsedList:
    """Outcome of :func:`parse_lei_list`."""

    leis: list[str] = field(default_factory=list)
    rejected: list[RejectedToken] = field(default_factory=list)
    #: Valid, unique LEIs beyond the cap that were NOT taken — so the UI can
    #: say "20 of 27 taken" instead of silently truncating.
    overflow: int = 0
    cap: int = MAX_ROWS

    @property
    def accepted(self) -> int:
        return len(self.leis)


def parse_lei_list(text: str, *, cap: int = MAX_ROWS) -> ParsedList:
    """Split pasted text into checksum-valid, de-duplicated LEIs.

    Every token is classified rather than the paste being refused whole:
    a wrong-length token, a shape miss, a check-digit failure and a
    duplicate each land in ``rejected`` with a reason a reader can act on.
    Order is preserved so the table can mirror the paste.
    """
    parsed = ParsedList(cap=cap)
    seen: set[str] = set()
    for raw in _SPLIT.split(text or ""):
        if not raw:
            continue
        lei = identifiers.normalise_lei(raw)
        if len(lei) != 20:
            parsed.rejected.append(
                RejectedToken(raw, f"{len(lei)} characters — an LEI has 20")
            )
            continue
        if not identifiers.LEI_STRICT_SHAPE.match(lei):
            parsed.rejected.append(
                RejectedToken(raw, "not an LEI: 18 letters or digits then two digits")
            )
            continue
        if identifiers.lei_check_digit_error(lei):
            parsed.rejected.append(
                RejectedToken(raw, "check digits do not match — a typo?")
            )
            continue
        if lei in seen:
            parsed.rejected.append(RejectedToken(raw, "duplicate"))
            continue
        seen.add(lei)
        if len(parsed.leis) >= cap:
            parsed.overflow += 1
            continue
        parsed.leis.append(lei)
    return parsed


def batch_concurrency() -> int:
    """Pipelines in flight per batch (``OPENCHECK_BATCH_CONCURRENCY``)."""
    return max(1, int(get_settings().batch_concurrency))


async def _one(
    lei: str, deepen_top: int, refresh: bool, sem: asyncio.Semaphore
) -> tuple[str, dict[str, Any], Any]:
    """Run one LEI through the single-lookup pipeline under the semaphore.

    Returns ``("row_done", row, response)`` or ``("row_failed", {...},
    None)``. Failures are *rows*, not exceptions: one unknown LEI must not
    abort the other nineteen, and a rate-limited anchor is reported so the
    reader knows that company was not checked. The full ``LookupResponse``
    rides along for callers that need more than the row — the combined
    export (Phase 167) needs every row's BODS statements.
    """
    from .mcp.shaping import shape_batch_row
    from .routers.lookup import _lookup_impl

    async with sem:
        try:
            resp = await _lookup_impl(lei, deepen_top=deepen_top, refresh=refresh)
        except HTTPException as exc:
            return (
                "row_failed",
                {
                    "lei": lei,
                    "status": exc.status_code,
                    "reason": str(exc.detail),
                    # 503 is the Phase 143/162 throttle refusal — momentary;
                    # everything else (404 unknown, 400 shape) is durable.
                    "retryable": exc.status_code == 503,
                    "degraded": True,
                },
                None,
            )
        except GleifRateLimitedError as exc:  # pragma: no cover — _lookup_impl maps it
            return (
                "row_failed",
                {
                    "lei": lei,
                    "status": 503,
                    "reason": str(exc) or "GLEIF is rate-limiting OpenCheck",
                    "retryable": True,
                    "degraded": True,
                },
                None,
            )
        except Exception as exc:  # noqa: BLE001 — a row, never a batch abort
            return (
                "row_failed",
                {
                    "lei": lei,
                    "status": 500,
                    "reason": f"{type(exc).__name__}: {exc}",
                    "retryable": False,
                    "degraded": True,
                },
                None,
            )
    return ("row_done", shape_batch_row(resp), resp)


async def iter_batch(
    leis: list[str],
    *,
    deepen_top: int = 5,
    refresh: bool = False,
    concurrency: int | None = None,
) -> AsyncIterator[tuple[str, dict[str, Any], Any]]:
    """Screen ``leis`` and yield ``(event, payload, response)`` per row.

    The engine under :func:`run_batch` (which drops the response) and the
    combined export (which keeps it). ``response`` is the ``LookupResponse``
    for ``row_done`` and ``None`` for ``row_failed`` and ``batch_done``.
    """
    sem = asyncio.Semaphore(concurrency or batch_concurrency())
    tasks = [
        asyncio.ensure_future(_one(lei, deepen_top, refresh, sem)) for lei in leis
    ]
    done = failed = 0
    try:
        for fut in asyncio.as_completed(tasks):
            event, payload, resp = await fut
            if event == "row_done":
                done += 1
            else:
                failed += 1
            yield (event, payload, resp)
    finally:
        # A client that disconnects mid-batch must not leave nineteen
        # pipelines running against the shared upstream budgets.
        for t in tasks:
            if not t.done():
                t.cancel()
    yield (
        "batch_done",
        {
            "requested": len(leis),
            "done": done,
            "failed": failed,
            # Rows that completed but with a screening check that did not
            # fully run are counted by the caller from each row's
            # ``degraded`` flag; ``failed`` rows are degraded by definition.
        },
        None,
    )


async def run_batch(
    leis: list[str],
    *,
    deepen_top: int = 5,
    refresh: bool = False,
    concurrency: int | None = None,
) -> AsyncIterator[tuple[str, dict[str, Any]]]:
    """Screen ``leis`` and yield ``(event, payload)`` as each row finishes.

    Events: ``row_done`` / ``row_failed`` per LEI (completion order), then
    one ``batch_done`` with the totals. Callers parse and cap the list
    first (:func:`parse_lei_list`) — this function trusts its input.
    """
    async for event, payload, _resp in iter_batch(
        leis, deepen_top=deepen_top, refresh=refresh, concurrency=concurrency
    ):
        yield (event, payload)


# ---------------------------------------------------------------------------
# Combined export (Phase 167)
# ---------------------------------------------------------------------------

#: The table as it is written to ``rows.csv`` — the same columns, in the
#: same order, as the page's client-side CSV (``frontend/src/lib/batch.ts``),
#: so a reader who has both files can line them up.
CSV_COLUMNS = [
    "lei",
    "legal_name",
    "jurisdiction",
    "register_status",
    "verdict",
    "risk_count",
    "risk_codes",
    "context_count",
    "context_codes",
    "sources_applicable",
    "sources_answered",
    "degraded",
    "degraded_sources",
    "state",
    "reason",
    "report_url",
]


@dataclass
class BatchResult:
    """Everything the combined export needs, in paste order."""

    rows: list[dict[str, Any]] = field(default_factory=list)
    failed: list[dict[str, Any]] = field(default_factory=list)
    #: Every row's BODS statements, de-duplicated by ``statementId`` across
    #: rows: two companies in one group share the parent's entity statement
    #: (both derive it from the same LEI), and it must appear once.
    statements: list[dict[str, Any]] = field(default_factory=list)
    #: Registered source ids that returned data for at least one row.
    contributing_ids: list[str] = field(default_factory=list)
    #: Per-bundle licence notices adapters attached, de-duplicated.
    license_notices: list[dict[str, Any]] = field(default_factory=list)
    #: ``statementId``s that appeared in more than one row's bundle.
    duplicate_statement_count: int = 0

    @property
    def totals(self) -> dict[str, int]:
        return {
            "requested": len(self.rows) + len(self.failed),
            "done": len(self.rows),
            "failed": len(self.failed),
            "degraded": sum(1 for r in self.rows if r.get("degraded")) + len(self.failed),
        }


async def collect_batch(
    leis: list[str], *, deepen_top: int = 5, refresh: bool = False
) -> BatchResult:
    """Run the batch to completion and merge every row's bundle.

    Rows come back in paste order regardless of completion order. A
    failed row contributes no statements and no sources — it is named in
    ``failed`` (and in ``rows.csv``) so the bundle never reads as a
    complete answer for a list it only partly covered.
    """
    order = {lei: i for i, lei in enumerate(leis)}
    rows: list[dict[str, Any]] = []
    failed: list[dict[str, Any]] = []
    responses: dict[str, Any] = {}
    async for event, payload, resp in iter_batch(
        leis, deepen_top=deepen_top, refresh=refresh
    ):
        if event == "row_done":
            rows.append(payload)
            responses[payload["lei"]] = resp
        elif event == "row_failed":
            failed.append(payload)
    rows.sort(key=lambda r: order.get(r["lei"], len(order)))
    failed.sort(key=lambda r: order.get(r["lei"], len(order)))

    result = BatchResult(rows=rows, failed=failed)
    seen: set[str] = set()
    sources: set[str] = set()
    notices: dict[str, dict[str, Any]] = {}
    for row in rows:
        resp = responses[row["lei"]]
        for stmt in resp.bods or []:
            sid = stmt.get("statementId")
            if sid and sid in seen:
                result.duplicate_statement_count += 1
                continue
            if sid:
                seen.add(sid)
            result.statements.append(stmt)
        for h in resp.hits or []:
            if not h.is_stub:
                sources.add(h.source_id)
        for n in resp.license_notices or []:
            key = json.dumps(n, sort_keys=True, default=str)
            notices.setdefault(key, n)
    result.contributing_ids = sorted(sources)
    result.license_notices = list(notices.values())
    return result


def _csv_row_done(row: dict[str, Any], origin: str) -> list[Any]:
    status = row.get("register_status") or {}
    cov = row.get("coverage") or {}
    return [
        row["lei"],
        row.get("legal_name") or "",
        row.get("jurisdiction") or "",
        status.get("liveness") or "",
        row.get("verdict") or "",
        row.get("risk_count", 0),
        " ".join(row.get("risk_codes") or []),
        row.get("context_count", 0),
        " ".join(row.get("context_codes") or []),
        cov.get("applicable", ""),
        cov.get("answered", ""),
        "true" if row.get("degraded") else "false",
        " ".join(row.get("degraded_sources") or []),
        "degraded" if row.get("degraded") else "done",
        "",
        f"{origin}{row.get('report_url', '')}",
    ]


def _csv_row_failed(row: dict[str, Any], origin: str) -> list[Any]:
    return [
        row["lei"], "", "", "", "", "", "", "", "", "", "", "true", "",
        "not checked", row.get("reason") or "", f"{origin}/?lei={row['lei']}",
    ]


def rows_csv(result: BatchResult, *, origin: str = "https://opencheck.world") -> str:
    """``rows.csv``: every screened LEI, one line each, failed rows included
    with their reason — the same columns the page's CSV button writes."""
    buf = io.StringIO()
    w = csv.writer(buf, lineterminator="\r\n")
    w.writerow(CSV_COLUMNS)
    for r in result.rows:
        w.writerow(_csv_row_done(r, origin))
    for r in result.failed:
        w.writerow(_csv_row_failed(r, origin))
    return buf.getvalue()

File: frontend/src/lib/batch.ts
/**
 * Batch screening — the page's decisions, kept out of the JSX (Phase 166).
 *
 * The frontend suite is logic-only, so every claim the `/batch` page makes
 * lives here where it can be pinned: how a paste is read, which rows sort
 * first, what the coverage cell says, what the cost line promises, and
 * what the CSV contains. `BatchPage.tsx` renders these and nothing else.
 *
 * Three rules from the ticket:
 *
 * - **Tolerant paste.** Newlines, commas, semicolons, spaces or tabs — a
 *   column copied out of a spreadsheet just works. Each token is checked
 *   for shape and ISO 17442 check digits *before* anything is sent, so the
 *   reader sees "check digits do not match" beside the token rather than
 *   an unknown-LEI row two minutes later. The backend re-checks; this is
 *   the same rule run early, never a substitute.
 * - **Twenty rows.** A cold anchor costs 4–6 GLEIF calls; twenty is about
 *   two minutes at the shared throttle. Anything beyond the cap is counted
 *   and said, never silently dropped.
 * - **A row that could not be checked is a row.** Failed and degraded rows
 *   sort to the top, because a list whose clean rows are read first is a
 *   list whose gaps are read last. Within a state, rows sort by risk count
 *   and then name — a sort, not a grade: OpenCheck does not rank companies
 *   by severity (Phase 132).
 */

import type { BatchFailedRow, BatchRow } from "./api";
import { coverageCopy, type CoverageCopy } from "./lookupProgress";

export const BATCH_CAP = 20;

/** Pipelines the backend runs at once — stated on the page, so keep it true. */
export const BATCH_CONCURRENCY = 2;

const SPLIT = /[\s,;]+/;
const SHAPE = /^[0-9A-Z]{18}[0-9]{2}$/;

export interface RejectedToken {
  token: string;
  reason: string;
}

export interface ParsedPaste {
  leis: string[];
  rejected: RejectedToken[];
  /** Valid, unique LEIs beyond the cap that were not taken. */
  overflow: number;
}

/**
 * ISO 17442 check digits: the LEI with letters as 10–35, digits as
 * themselves, mod 97 must be 1. The backend's `lei_check_digits_ok`,
 * in TypeScript, so a typo is caught before a request is spent on it.
 */
export function leiCheckDigitsOk(lei: string): boolean {
  if (!SHAPE.test(lei)) return false;
  let remainder = 0;
  for (const ch of lei) {
    const value = ch >= "A" && ch <= "Z" ? ch.charCodeAt(0) - 55 : Number(ch);
    // Feed the value's decimal digits one at a time — letters contribute two.
    for (const d of String(value)) remainder = (remainder * 10 + Number(d)) % 97;
  }
  return remainder === 1;
}

export function parseLeiPaste(text: string, cap: number = BATCH_CAP): ParsedPaste {
  const out: ParsedPaste = { leis: [], rejected: [], overflow: 0 };
  const seen = new Set<string>();
  for (const raw of (text || "").split(SPLIT)) {
    if (!raw) continue;
    const lei = raw.trim().toUpperCase();
    if (lei.length !== 20) {
      out.rejected.push({ token: raw, reason: `${lei.length} characters — an LEI has 20` });
      continue;
    }
    if (!SHAPE.test(lei)) {
      out.rejected.push({ token: raw, reason: "not an LEI: 18 letters or digits then two digits" });
      continue;
    }
    if (!leiCheckDigitsOk(lei)) {
      out.rejected.push({ token: raw, reason: "check digits do not match — a typo?" });
      continue;
    }
    if (seen.has(lei)) {
      out.rejected.push({ token: raw, reason: "duplicate" });
      continue;
    }
    seen.add(lei);
    if (out.leis.length >= cap) {
      out.overflow += 1;
      continue;
    }
    out.leis.push(lei);
  }
  return out;
}

/** "3 valid · 1 rejected · 2 beyond the cap of 20" — the live count under the box. */
export function pasteSummary(p: ParsedPaste, cap: number = BATCH_CAP): string {
  const parts = [`${p.leis.length} valid`];
  if (p.rejected.length) parts.push(`${p.rejected.length} rejected`);
  if (p.overflow) parts.push(`${p.overflow} beyond the cap of ${cap}`);
  return parts.join(" · ");
}

/**
 * What the run will cost, said before the button is pressed. Two pipelines
 * at a time against a shared upstream budget; a company seen in the last
 * fifteen minutes is free, so this is the ceiling, not a promise.
 */
export function costLine(n: number, concurrency: number = BATCH_CONCURRENCY): string {
  if (n <= 0) return "";
  const company = n === 1 ? "company" : "companies";
  if (n === 1) return "A few seconds for a company not seen recently.";
  // ~6 s per cold anchor per pipeline (4–6 GLEIF calls at 50/min, shared).
  const seconds = Math.ceil(n / concurrency) * 6;
  const time =
    seconds < 45
      ? "under a minute"
      : seconds < 90
        ? "about a minute"
        : `about ${Math.round(seconds / 60)} minutes`;
  return `${capitalise(time)} for ${n} ${company} not seen recently — OpenCheck runs ${concurrency} at a time and queues behind GLEIF's rate limit rather than tripping it.`;
}

function capitalise(s: string): string {
  return s.charAt(0).toUpperCase() + s.slice(1);
}

/** The three states a row can be in on the page. */
export type RowState = "running" | "done" | "degraded";

export type TableRow =
  | { state: "running"; lei: string }
  | { state: "done" | "degraded"; lei: string; row: BatchRow }
  | { state: "degraded"; lei: string; failed: BatchFailedRow };

export function rowState(row: BatchRow): "done" | "degraded" {
  return row.degraded ? "degraded" : "done";
}

/**
 * Companies that could not be checked at all first, then rows whose check
 * did not fully run, then done rows by risk count descending and name.
 * Running rows keep paste order at the end so the table does not reshuffle
 * under the reader while the stream is open.
 */
export function sortRows(rows: TableRow[]): TableRow[] {
  const rank = (r: TableRow): number =>
    "failed" in r ? 0 : r.state === "degraded" ? 1 : r.state === "done" ? 2 : 3;
  const risk = (r: TableRow): number => ("row" in r ? r.row.risk_count : -1);
  const name = (r: TableRow): string => ("row" in r ? (r.row.legal_name ?? "") : "");
  return rows
    .map((r, i) => ({ r, i }))
    .sort((a, b) => {
      const d = rank(a.r) - rank(b.r);
      if (d) return d;
      const rd = risk(b.r) - risk(a.r);
      if (rd) return rd;
      const nd = name(a.r).localeCompare(name(b.r));
      if (nd) return nd;
      return a.i - b.i;
    })
    .map(({ r }) => r);
}

/**
 * The coverage cell, from the same `coverageCopy` the single report uses
 * (Phase 156). The row counts the GLEIF anchor in both figures already, so
 * it is taken back out before the helper adds it.
 */
export function rowCoverage(row: BatchRow, registryTotal: number | null): CoverageCopy {
  return coverageCopy({
    answered: Math.max(0, row.coverage.answered - 1),
    applicable: Math.max(0, row.coverage.applicable - 1),
    total: registryTotal,
    jurisdiction: row.jurisdiction,
    screening: false,
    anchorAnswered: true,
  });
}

const CSV_HEADER = [
  "lei",
  "legal_name",
  "jurisdiction",
  "register_status",
  "verdict",
  "risk_count",
  "risk_codes",
  "context_count",
  "context_codes",
  "sources_applicable",
  "sources_answered",
  "degraded",
  "degraded_sources",
  "state",
  "reason",
  "report_url",
];

function csvCell(v: unknown): string {
  const s = v === null || v === undefined ? "" : String(v);
  return /[",\n\r]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

/**
 * The table as CSV — every screened LEI, one line each, failed rows
 * included with their reason so the file never reads as a clean list of
 * the ones that worked. `report_url` is absolute when `origin` is given.
 */
export function rowsToCsv(rows: TableRow[], origin = ""): string {
  const lines = [CSV_HEADER.join(",")];
  for (const r of rows) {
    if (r.state === "running") continue;
    if ("failed" in r) {
      lines.push(
        [
          r.lei, "", "", "", "", "", "", "", "", "", "", "true", "", "not checked",
          r.failed.reason, `${origin}/?lei=${r.lei}`,
        ].map(csvCell).join(","),
      );
      continue;
    }
    const row = r.row;
    lines.push(
      [
        row.lei,
        row.legal_name ?? "",
        row.jurisdiction ?? "",
        row.register_status?.liveness ?? "",
        row.verdict ?? "",
        row.risk_count,
        row.risk_codes.join(" "),
        row.context_count,
        row.context_codes.join(" "),
        row.coverage.applicable,
        row.coverage.answered,
        row.degraded ? "true" : "false",
        row.degraded_sources.join(" "),
        row.degraded ? "degraded" : "done",
        "",
        `${origin}${row.report_url}`,
      ].map(csvCell).join(","),
    );
  }
  return lines.join("\r\n") + "\r\n";
}

/** `opencheck-batch-2026-09-03.csv` */
export function csvFilename(now: Date = new Date()): string {
  return `opencheck-batch-${now.toISOString().slice(0, 10)}.csv`;
}
